GVFS/GVFS.UnitTests/Common/GVFSLockTests.cs

```diff
@@ -0,0 +1,206 @@
+        }
+
+        [TestCase]
+        [Category(CategoryConstants.ExceptionExpected)]
+        public void ReleaseLockHeldByGVFS_WhenNoLock()
+        {
+            GVFSLock gvfsLock = new GVFSLock(new MockTracer());
+            this.ValidateLockIsFree(gvfsLock);
+            Assert.Throws<InvalidOperationException>(() => gvfsLock.ReleaseLockHeldByGVFS());
+        }
+
+        [TestCase]
+        [Category(CategoryConstants.ExceptionExpected)]
+        public void ReleaseLockHeldByGVFS_WhenExternalHasLockShouldThrow()
+        {
+            MockPlatform mockPlatform = (MockPlatform)GVFSPlatform.Instance;
+            GVFSLock gvfsLock = this.AcquireDefaultLock(mockPlatform);
+
+            Assert.Throws<InvalidOperationException>(() => gvfsLock.ReleaseLockHeldByGVFS());
+        }
+
+        [TestCase]
+        public void TryAcquireLockForGVFS()
+        {
+            GVFSLock gvfsLock = this.AcquireGVFSLock();
+
+            // Should be able to call again when GVFS has the lock
+            gvfsLock.TryAcquireLockForGVFS().ShouldBeTrue();
+            this.ValidateLockHeldByGVFS(gvfsLock);
+
+            gvfsLock.ReleaseLockHeldByGVFS();
+            this.ValidateLockIsFree(gvfsLock);
+        }
+
+        [TestCase]
+        public void TryAcquireLockForGVFS_WhenExternalLock()
+        {
+            MockPlatform mockPlatform = (MockPlatform)GVFSPlatform.Instance;
+            GVFSLock gvfsLock = this.AcquireDefaultLock(mockPlatform);
+
+            gvfsLock.TryAcquireLockForGVFS().ShouldBeFalse();
+            mockPlatform.ActiveProcesses.Remove(DefaultLockData.PID);
+        }
+
+        [TestCase]
+        public void TryAcquireLockForExternalRequestor_WhenGVFSLock()
+        {
+            GVFSLock gvfsLock = this.AcquireGVFSLock();
+
+            NamedPipeMessages.LockData existingExternalHolder;
+            gvfsLock.TryAcquireLockForExternalRequestor(DefaultLockData, out existingExternalHolder).ShouldBeFalse();
+            this.ValidateLockHeldByGVFS(gvfsLock);
+            existingExternalHolder.ShouldBeNull();
+        }
+
+        [TestCase]
+        public void TryAcquireLockForExternalRequestor_WhenExternalLock()
+        {
+            MockPlatform mockPlatform = (MockPlatform)GVFSPlatform.Instance;
+            GVFSLock gvfsLock = this.AcquireDefaultLock(mockPlatform);
+
+            NamedPipeMessages.LockData newLockData = new NamedPipeMessages.LockData(4321, false, false, "git new");
+            NamedPipeMessages.LockData existingExternalHolder;
+            gvfsLock.TryAcquireLockForExternalRequestor(newLockData, out existingExternalHolder).ShouldBeFalse();
+            this.ValidateLockHeld(gvfsLock, DefaultLockData);
+            this.ValidateExistingExternalHolder(DefaultLockData, existingExternalHolder);
+            mockPlatform.ActiveProcesses.Remove(DefaultLockData.PID);
+        }
+
+        [TestCase]
+        public void AbleToAcquireLockAfterTerminatedProcess()

```

Comment:
Minor naming suggestion:  Something like `TryAcquireLockForExternalRequestor_WhenExternalHolderTerminated` would better match the other `TryAcquireLockForExternalRequestor` tests.